database/raft/raft.go

```diff
@@ -83,14 +81,28 @@ type Service struct {
 	// The actual replicated data set.
 	stateMu   sync.Mutex
 	stateCond sync.Cond
-	state     *state.State
+	state     State
 	confState raftpb.ConfState
 	done      bool
 
 	// Current log position, accessed only from runUpdates goroutine
 	snapIndex uint64
 }
 
+type State interface {
```

Comment:
to circumvent a bidirectional dependency. I was hoping that database/raft would only be raft-specific logic and completely ignorant of its use in implementing a kv store, but it needs a way to access and modify the raft state. This interface keeps key-value logic out of the raft package but still lets it interact with the key-value state.